pyccel/ast/basic.py

```diff
@@ -442,3 +442,19 @@ def copy_attributes(self, x):
         self._precision = x.precision
         self._order     = x.order
 
+
+#------------------------------------------------------------------------------
+class ScopedNode(Basic):
+    """ Class from which all objects with a scope inherit
+    """
+    __slots__ = ('_scope',)
+
+    def __init__(self, scope = None):
+        self._scope = scope
+        super().__init__()
+
+    @property
+    def scope(self):
+        """ Scope object containing all available objects in this part of the code
+        """
+        return self._scope
```

Comment:
A `ScopedNode` is a node which has a scope. I.e. which has variables, functions, etc which are accessible within it. This is not particularly related to `OpenMP` variables. In Python a `While` would not be a `ScopedNode`, but it can be in other languages where temporaries can be created inside a loop which is why a scope is created for it.

This is just a way to associate the existing scopes stored in `Parser.namespace` with the ast node that describes the scope.

As written when in a function `FunctionDef.scope` (as `FunctionDef` is a `ScopedNode`) corresponds to the **local** scope, `FunctionDef.scope.parent_scope` is the **enclosing** scope. No distinction is made between the **global** scope and the outermost **enclosing** scope

Reply:
Thanks for the explanation! How about the `scope = None` case?

Reply:
Currently (the next PR will change this), there are no scopes at the syntactic stage. So at this point `scope = None`

Reply:
It is also useful for some corner cases such as `PyccelFunctionDef` which inherits from `FunctionDef` but doesn't need a `Scope`